File: backend/app/api/floor.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.base import get_db
from app.models.floor import Floor
from app.schemas.floor import FloorCreate, FloorResponse, FloorUpdate
from app.services.floor_service import FloorService

router = APIRouter(tags=["Floors"])


@router.post("/floors", response_model=FloorResponse, status_code=201)
async def create_floor(data: FloorCreate, db: AsyncSession = Depends(get_db)) -> Floor:
    service = FloorService(db)
    return await service.create_floor(data)


@router.get("/floors", response_model=list[FloorResponse])
async def list_floors(db: AsyncSession = Depends(get_db)) -> list[Floor]:
    service = FloorService(db)
    return await service.get_all_floors()


@router.get("/floors/{floor_id}", response_model=FloorResponse)
async def get_floor(floor_id: int, db: AsyncSession = Depends(get_db)) -> Floor:
    service = FloorService(db)
    return await service.get_floor(floor_id)


@router.put("/floors/{floor_id}", response_model=FloorResponse)
async def update_floor(
    floor_id: int, data: FloorUpdate, db: AsyncSession = Depends(get_db)
) -> Floor:
    service = FloorService(db)
    return await service.update_floor(floor_id, data)


@router.delete("/floors/{floor_id}", status_code=204)
async def delete_floor(floor_id: int, db: AsyncSession = Depends(get_db)) -> None:
    service = FloorService(db)
    await service.delete_floor(floor_id)


@router.get("/buildings/{building_id}/floors", response_model=list[FloorResponse])
async def get_floors_by_building(
    building_id: int, db: AsyncSession = Depends(get_db)
) -> list[Floor]:
    service = FloorService(db)
    return await service.get_floors_by_building(building_id)

File: frontend/src/services/floor.ts
import { api } from "@/services/api";
import type { Floor, FloorCreateInput, FloorUpdateInput } from "@/types/floor";

export const floorService = {
  list: async (): Promise<Floor[]> => {
    const { data } = await api.get<Floor[]>("/floors");
    return data;
  },

  get: async (id: number): Promise<Floor> => {
    const { data } = await api.get<Floor>(`/floors/${id}`);
    return data;
  },

  listByBuilding: async (buildingId: number): Promise<Floor[]> => {
    const { data } = await api.get<Floor[]>(`/buildings/${buildingId}/floors`);
    return data;
  },

  create: async (payload: FloorCreateInput): Promise<Floor> => {
    const { data } = await api.post<Floor>("/floors", payload);
    return data;
  },

  update: async (id: number, payload: FloorUpdateInput): Promise<Floor> => {
    const { data } = await api.put<Floor>(`/floors/${id}`, payload);
    return data;
  },

  remove: async (id: number): Promise<void> => {
    await api.delete(`/floors/${id}`);
  },
};
